File: server/src/sync/sync.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import {
  PullResponseDto,
  HabitDto,
  CheckinDto,
  TombstoneDto
} from './dto/pull-response.dto';
import {
  PushConflict,
  PushResponseDto,
  SyncOpDto
} from './dto/push-request.dto';
import { MetricsService } from '../metrics/metrics.service';

interface Cursor {
  updatedAt: Date;
  id: string;
}

interface HabitPayload {
  id: string;
  name: string;
  description?: string | null;
  color: string;
  icon: string;
  frequency: string;
  targetStreak: number;
  tags?: unknown;
  archived?: boolean;
  version?: number;
  updatedAt?: string;
  createdAt?: string;
}

interface CheckinPayload {
  id?: string;
  habitId: string;
  date: string;
  done: boolean;
  version?: number;
  updatedAt?: string;
}

type TxClient = Omit<
  PrismaClient,
  '$on' | '$connect' | '$disconnect' | '$use' | '$transaction' | '$extends'
>;

@Injectable()
export class SyncService {
  private readonly logger = new Logger(SyncService.name);

  constructor(
    private readonly prisma: PrismaService,
    private readonly metrics: MetricsService
  ) {}

  async pull(userId: string, since?: string, traceId?: string): Promise<PullResponseDto> {
    const pullStart = Date.now();
    try {
      const cursor = this.parseCursor(since);
      const updatedFilter = cursor
        ? { AND: [this.buildCursorClause(cursor, 'updatedAt')] }
        : undefined;
      const deletedFilter = cursor
        ? { AND: [this.buildCursorClause(cursor, 'deletedAt')] }
        : undefined;

      const habits = await this.prisma.habit.findMany({
        where: {
          userId,
          ...updatedFilter
        },
        orderBy: [
          { updatedAt: 'asc' },
          { id: 'asc' }
        ],
        take: 200
      });

      const checkins = await this.prisma.checkin.findMany({
        where: {
          userId,
          ...updatedFilter
        },
        orderBy: [
          { updatedAt: 'asc' },
          { id: 'asc' }
        ],
        take: 200
      });

      const tombstones = await this.prisma.tombstone.findMany({
        where: {
          userId,
          ...deletedFilter
        },
        orderBy: [
          { deletedAt: 'asc' },
          { id: 'asc' }
        ],
        take: 200
      });

      const cursorCandidates = [
        ...habits.map((h: { updatedAt: Date; id: string }) => ({ updatedAt: h.updatedAt, id: h.id })),
        ...checkins.map((c: { updatedAt: Date; id: string }) => ({ updatedAt: c.updatedAt, id: c.id })),
        ...tombstones.map((t: { deletedAt: Date; id: string }) => ({ updatedAt: t.deletedAt, id: t.id }))
      ];

      const nextCursor = this.calculateNextCursor(cursorCandidates);
      const serverTime = new Date().toISOString();

      this.metrics.recordPull(
        Date.now() - pullStart,
        habits.length + checkins.length + tombstones.length
      );

      return {
        habits: habits.map(this.serializeHabit),
        checkins: checkins.map(this.serializeCheckin),
        tombstones: tombstones.map(this.serializeTombstone),
        nextCursor,
        serverTime
      };
    } catch (error) {
      this.metrics.recordError();
      this.logger.error(
        `sync pull failed userId=${userId} traceId=${traceId ?? 'n/a'}`,
        error instanceof Error ? error.stack : String(error)
      );
      throw error;
    }
  }

  async push(userId: string, ops: SyncOpDto[], traceId?: string): Promise<PushResponseDto> {
    const pushStart = Date.now();
    const applied: string[] = [];
    const conflicts: PushConflict[] = [];
    const serverTime = new Date().toISOString();

    try {
      await this.prisma.$transaction(async (tx: TxClient) => {
        for (const op of ops) {
          if (!op.id) continue;

          const deduplicated = await this.tryCreateLog(tx, op.id);
          if (!deduplicated) continue;

          if (op.entity === 'habit') {
            await this.applyHabitOp(tx, userId, op, applied, conflicts);
          } else if (op.entity === 'checkin') {
            await this.applyCheckinOp(tx, userId, op, applied, conflicts);
          }
        }
      });
    } catch (error) {
      this.metrics.recordError();
      this.logger.error(
        `sync push failed userId=${userId} traceId=${traceId ?? 'n/a'}`,
        error instanceof Error ? error.stack : String(error)
      );
      throw error;
    }

    this.metrics.recordPush(Date.now() - pushStart, conflicts.length, ops.length);
    return { applied, conflicts, serverTime };
  }

  private async applyHabitOp(
    tx: TxClient,
    userId: string,
    op: SyncOpDto,
    applied: string[],
    conflicts: PushConflict[]
  ) {
    const payload = op.payload as unknown as HabitPayload;
    if (!payload?.id) return;
    const timestamp = this.normalizeDate(payload.updatedAt);

    if (op.type === 'delete') {
      await tx.tombstone.create({
        data: {
          userId,
          entity: 'habit',
          entityId: payload.id,
          version: payload.version ?? 1
        }
      });
      await tx.habit.deleteMany({ where: { id: payload.id, userId } });
      applied.push(op.id);
      return;
    }

    const existing = await tx.habit.findUnique({ where: { id: payload.id } });
    if (existing && existing.userId !== userId) {
      conflicts.push({
        opId: op.id,
        reason: 'habit belongs to another user'
      });
      return;
    }
    if (
      existing &&
      new Date(existing.updatedAt).getTime() > timestamp.getTime()
    ) {
      conflicts.push({
        opId: op.id,
        reason: 'server already has newer habit',
        serverValue: {
          version: existing.version,
          updatedAt: existing.updatedAt
        }
      });
      return;
    }

    const nextVersion = Math.max(existing?.version ?? 0, payload.version ?? 0) + 1;

    const tags = this.normalizeTags(payload.tags);

    await tx.habit.upsert({
      where: { id: payload.id },
      create: {
        id: payload.id,
        userId,
        name: payload.name,
        description: payload.description ?? null,
        color: payload.color,
        icon: payload.icon,
        frequency: payload.frequency,
        targetStreak: payload.targetStreak,
        tags: tags as never,
        archived: payload.archived ?? false,
        createdAt: this.normalizeDate(payload.createdAt),
        updatedAt: timestamp,
        version: nextVersion
      },
      update: {
        name: payload.name,
        description: payload.description ?? null,
        color: payload.color,
        icon: payload.icon,
        frequency: payload.frequency,
        targetStreak: payload.targetStreak,
        tags: tags as never,
        archived: payload.archived ?? false,
        updatedAt: timestamp,
        version: nextVersion
      }
    });

    applied.push(op.id);
  }

  private async applyCheckinOp(
    tx: TxClient,
    userId: string,
    op: SyncOpDto,
    applied: string[],
    conflicts: PushConflict[]
  ) {
    const payload = op.payload as unknown as CheckinPayload;
    if (!payload?.habitId || !payload.date) return;
    const timestamp = this.normalizeDate(payload.updatedAt);
    const date = new Date(payload.date);
    const parentHabit = await tx.habit.findUnique({
      where: { id: payload.habitId },
      select: { userId: true }
    });
    if (!parentHabit || parentHabit.userId !== userId) {
      conflicts.push({
        opId: op.id,
        reason: 'checkin habit belongs to another user'
      });
      return;
    }

    if (op.type === 'delete') {
      await tx.tombstone.create({
        data: {
          userId,
          entity: 'checkin',
          entityId: payload.id ?? `${payload.habitId}:${payload.date}`,
          version: payload.version ?? 1
        }
      });
      await tx.checkin.deleteMany({
        where: { habitId: payload.habitId, date, userId }
      });
      applied.push(op.id);
      return;
    }

    const existing = await tx.checkin.findFirst({
      where: { habitId: payload.habitId, date, userId }
    });
    if (
      existing &&
      new Date(existing.updatedAt).getTime() > timestamp.getTime()
    ) {
      conflicts.push({
        opId: op.id,
        reason: 'server already has newer checkin',
        serverValue: {
          version: existing.version,
          updatedAt: existing.updatedAt
        }
      });
      return;
    }

    const nextVersion =
      Math.max(existing?.version ?? 0, payload.version ?? 0) + 1;

    await tx.checkin.upsert({
      where: {
        habit_date_unique: {
          habitId: payload.habitId,
          date
        }
      },
      create: {
        habitId: payload.habitId,
        userId,
        date,
        done: payload.done,
        updatedAt: timestamp,
        version: nextVersion
      },
      update: {
        done: payload.done,
        updatedAt: timestamp,
        version: nextVersion
      }
    });

    applied.push(op.id);
  }

  private async tryCreateLog(
    tx: TxClient,
    opId: string
  ): Promise<boolean> {
    try {
      await tx.syncOpLog.create({ data: { opId } });
      return true;
    } catch (error) {
      if (this.isUniqueConstraintError(error)) {
        return false;
      }
      throw error;
    }
  }

  private parseCursor(cursor?: string): Cursor | undefined {
    if (!cursor) return undefined;
    const parts = cursor.split('|');
    if (parts.length < 2) return undefined;
    const date = new Date(parts[0]);
    const id = parts[1];
    if (Number.isNaN(date.getTime())) return undefined;
    return { updatedAt: date, id };
  }

  private buildCursorClause(
    cursor: Cursor,
    field: 'updatedAt' | 'deletedAt'
  ) {
    return {
      OR: [
        { [field]: { gt: cursor.updatedAt } },
        {
          [field]: { equals: cursor.updatedAt },
          id: { gt: cursor.id }
        }
      ]
    };
  }

  private calculateNextCursor(
    records: { updatedAt: Date; id: string }[]
  ): string | undefined {
    if (records.length === 0) return undefined;
    const sorted = records.sort((a, b) => {
      const diff = a.updatedAt.getTime() - b.updatedAt.getTime();
      if (diff !== 0) return diff;
      return a.id.localeCompare(b.id);
    });
    const last = sorted[sorted.length - 1];
    return `${last.updatedAt.toISOString()}|${last.id}`;
  }

  private serializeHabit(habit: {
    id: string;
    name: string;
    description: string | null;
    color: string;
    icon: string;
    frequency: string;
    targetStreak: number;
    tags: unknown;
    archived: boolean;
    createdAt: Date;
    updatedAt: Date;
    version: number;
  }): HabitDto {
    return {
      id: habit.id,
      name: habit.name,
      description: habit.description,
      color: habit.color,
      icon: habit.icon,
      frequency: habit.frequency,
      targetStreak: habit.targetStreak,
      tags: habit.tags ?? undefined,
      archived: habit.archived,
      createdAt: habit.createdAt.toISOString(),
      updatedAt: habit.updatedAt.toISOString(),
      version: habit.version
    };
  }

  private serializeCheckin(checkin: {
    id: string;
    habitId: string;
    date: Date;
    done: boolean;
    updatedAt: Date;
    version: number;
  }): CheckinDto {
    return {
      id: checkin.id,
      habitId: checkin.habitId,
      date: checkin.date.toISOString(),
      done: checkin.done,
      updatedAt: checkin.updatedAt.toISOString(),
      version: checkin.version
    };
  }

  private serializeTombstone(tombstone: {
    id: string;
    entity: string;
    entityId: string;
    deletedAt: Date;
    version: number;
  }): TombstoneDto {
    return {
      id: tombstone.id,
      entity: tombstone.entity,
      entityId: tombstone.entityId,
      deletedAt: tombstone.deletedAt.toISOString(),
      version: tombstone.version
    };
  }

  private normalizeDate(value?: string): Date {
    if (!value) return new Date();
    const parsed = new Date(value);
    if (Number.isNaN(parsed.getTime())) return new Date();
    return parsed;
  }

  private isUniqueConstraintError(error: unknown): boolean {
    if (!error || typeof error !== 'object') return false;
    if (!('code' in error)) return false;
    return (error as { code?: string }).code === 'P2002';
  }

  private normalizeTags(value: unknown): unknown {
    if (value === undefined) return undefined;
    if (value === null) return undefined;
    return value;
  }
}

File: apps/backend/src/main/java/com/habittracker/sync/SyncService.java
package com.habittracker.sync;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.habittracker.model.CheckinEntity;
import com.habittracker.model.HabitEntity;
import com.habittracker.model.TombstoneEntity;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.persistence.PersistenceException;
import jakarta.transaction.Transactional;
import com.habittracker.model.SyncOpLogEntity;

import java.math.BigInteger;
import java.time.Instant;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;

@ApplicationScoped
public class SyncService {
  private static final DateTimeFormatter ISO = DateTimeFormatter.ISO_OFFSET_DATE_TIME;
  private static final TypeReference<Map<String, Object>> MAP_TYPE = new TypeReference<>() {
  };

  final ObjectMapper objectMapper;

  public SyncService(ObjectMapper objectMapper) {
    this.objectMapper = objectMapper;
  }

  // ─── Pull ─────────────────────────────────────────────────────────────────

  public SyncDtos.PullResponseDto pull(String userId, String since) {
    var cursor = parseCursor(since);

    List<HabitEntity> habits;
    List<CheckinEntity> checkins;
    List<TombstoneEntity> tombstones;

    if (cursor == null) {
      habits = HabitEntity.<HabitEntity>find(
          "userId = ?1 ORDER BY updatedAt ASC, id ASC", userId
      ).page(0, 200).list();
      checkins = CheckinEntity.<CheckinEntity>find(
          "userId = ?1 ORDER BY updatedAt ASC, id ASC", userId
      ).page(0, 200).list();
      tombstones = TombstoneEntity.<TombstoneEntity>find(
          "userId = ?1 ORDER BY deletedAt ASC, id ASC", userId
      ).page(0, 200).list();
    } else {
      habits = HabitEntity.<HabitEntity>find(
          "userId = ?1 AND (updatedAt > ?2 OR (updatedAt = ?2 AND id > ?3)) ORDER BY updatedAt ASC, id ASC",
          userId, cursor.updatedAt(), cursor.id()
      ).page(0, 200).list();
      checkins = CheckinEntity.<CheckinEntity>find(
          "userId = ?1 AND (updatedAt > ?2 OR (updatedAt = ?2 AND id > ?3)) ORDER BY updatedAt ASC, id ASC",
          userId, cursor.updatedAt(), cursor.id()
      ).page(0, 200).list();
      tombstones = TombstoneEntity.<TombstoneEntity>find(
          "userId = ?1 AND (deletedAt > ?2 OR (deletedAt = ?2 AND id > ?3)) ORDER BY deletedAt ASC, id ASC",
          userId, cursor.updatedAt(), cursor.id()
      ).page(0, 200).list();
    }

    var candidates = new ArrayList<CursorRow>();
    habits.forEach(h -> candidates.add(new CursorRow(h.updatedAt, h.id)));
    checkins.forEach(c -> candidates.add(new CursorRow(c.updatedAt, c.id)));
    tombstones.forEach(t -> candidates.add(new CursorRow(t.deletedAt, t.id)));

    var nextCursor = calculateNextCursor(candidates);

    return new SyncDtos.PullResponseDto(
        habits.stream().map(this::serializeHabit).toList(),
        checkins.stream().map(this::serializeCheckin).toList(),
        tombstones.stream().map(this::serializeTombstone).toList(),
        nextCursor,
        toSyncIso(Instant.now())
    );
  }

  // ─── Push ─────────────────────────────────────────────────────────────────

  @Transactional
  public SyncDtos.PushResponseDto push(String userId, List<SyncDtos.SyncOpDto> ops) {
    var applied = new ArrayList<String>();
    var conflicts = new ArrayList<SyncDtos.PushConflict>();

    for (var op : ops) {
      if (op.id() == null || op.id().isBlank()) {
        continue;
      }
      if (!tryCreateLog(op.id())) {
        continue;
      }
      if ("habit".equals(op.entity())) {
        applyHabitOp(userId, op, applied, conflicts);
      } else if ("checkin".equals(op.entity())) {
        applyCheckinOp(userId, op, applied, conflicts);
      }
    }

    return new SyncDtos.PushResponseDto(applied, conflicts, toSyncIso(Instant.now()));
  }

  // ─── Habit ops ────────────────────────────────────────────────────────────

  private void applyHabitOp(
      String userId,
      SyncDtos.SyncOpDto op,
      List<String> applied,
      List<SyncDtos.PushConflict> conflicts
  ) {
    var payload = toMap(op.payload());
    var habitId = asString(payload.get("id"));
    if (habitId == null) {
      return;
    }

    if ("delete".equals(op.type())) {
      deleteHabit(userId, habitId, payload);
      applied.add(op.id());
      return;
    }

    var existing = (HabitEntity) HabitEntity.findById(habitId);
    if (existing != null && !userId.equals(existing.userId)) {
      conflicts.add(new SyncDtos.PushConflict(op.id(), "habit belongs to another user", null));
      return;
    }

    var clientUpdated = normalizeInstant(asString(payload.get("updatedAt")));
    if (existing != null && existing.updatedAt.isAfter(clientUpdated)) {
      conflicts.add(new SyncDtos.PushConflict(op.id(), "server already has newer habit", Map.of(
          "version", existing.version,
          "updatedAt", toSyncIso(existing.updatedAt)
      )));
      return;
    }

    boolean isNew = false;
    if (existing == null) {
      existing = new HabitEntity();
      existing.id = habitId;
      existing.userId = userId;
      var raw = asString(payload.get("createdAt"));
      existing.createdAt = raw != null ? parseInstantOrNow(raw) : Instant.now();
      isNew = true;
    }

    var name = asString(payload.get("name"));
    existing.name = name != null ? name : (existing.name != null ? existing.name : "Habit");
    existing.description = nullableString(payload.get("description"));
    var color = asString(payload.get("color"));
    existing.color = color != null ? color : (existing.color != null ? existing.color : "#5E81AC");
    var icon = asString(payload.get("icon"));
    existing.icon = icon != null ? icon : (existing.icon != null ? existing.icon : "star");
    var frequency = asString(payload.get("frequency"));
    existing.frequency = frequency != null ? frequency : (existing.frequency != null ? existing.frequency : "daily");
    existing.customDays = normalizeCustomDaysJson(payload.get("customDays"));
    existing.schedule = jsonOrNull(payload.get("schedule"));
    existing.targetStreak = asInt(payload.get("targetStreak"), 1);
    existing.dailyTarget = resolveDailyTarget(payload.get("dailyTarget"), existing.dailyTarget);
    existing.tags = jsonOrNull(payload.get("tags"));
    existing.archived = asBoolean(payload.get("archived"), false);
    existing.sortOrder = resolveSortOrder(payload.get("sortOrder"), existing.sortOrder);
    existing.reminderTime = normalizeReminderTime(asString(payload.get("reminderTime")));
    existing.reminderEnabled = asBoolean(payload.get("reminderEnabled"), existing.reminderEnabled);
    existing.type = normalizeType(asString(payload.get("type")));
    existing.freezeDays = normalizeFreezeDaysJson(payload.get("freezeDays"), existing.freezeDays);
    existing.version = Math.max(existing.version, asInt(payload.get("version"), 0)) + 1;
    existing.updatedAt = nextSyncDate(clientUpdated, existing.updatedAt);

    if (isNew) {
      existing.persist();
    }

    applied.add(op.id());
  }

  private void deleteHabit(String userId, String habitId, Map<String, Object> payload) {
    var tombstone = new TombstoneEntity();
    tombstone.userId = userId;
    tombstone.entity = "habit";
    tombstone.entityId = habitId;
    tombstone.version = asInt(payload.get("version"), 1);
    tombstone.deletedAt = nextSyncDate(parseInstantOrNow(asString(payload.get("updatedAt"))));
    tombstone.persist();

    CheckinEntity.delete("habitId = ?1 and userId = ?2", habitId, userId);
    HabitEntity.delete("id = ?1 and userId = ?2", habitId, userId);
  }

  // ─── Checkin ops ──────────────────────────────────────────────────────────

  private void applyCheckinOp(
      String userId,
      SyncDtos.SyncOpDto op,
      List<String> applied,
      List<SyncDtos.PushConflict> conflicts
  ) {
    var payload = toMap(op.payload());
    var habitId = asString(payload.get("habitId"));
    var dateString = asString(payload.get("date"));
    if (habitId == null || dateString == null) {
      return;
    }

    var parent = (HabitEntity) HabitEntity.findById(habitId);
    if (parent == null || !userId.equals(parent.userId)) {
      conflicts.add(new SyncDtos.PushConflict(op.id(), "checkin habit belongs to another user", null));
      return;
    }

    var date = toLocalDate(dateString);
    var existing = CheckinEntity.<CheckinEntity>find(
        "habitId = ?1 and date = ?2 and userId = ?3", habitId, date, userId
    ).firstResult();
    var clientUpdated = normalizeInstant(asString(payload.get("updatedAt")));

    if ("delete".equals(op.type())) {
      deleteCheckin(userId, habitId, date, payload, existing);
      applied.add(op.id());
      return;
    }

    if (existing != null && existing.updatedAt.isAfter(clientUpdated)) {
      conflicts.add(new SyncDtos.PushConflict(op.id(), "server already has newer checkin", Map.of(
          "version", existing.version,
          "updatedAt", toSyncIso(existing.updatedAt)
      )));
      return;
    }

    if (existing == null) {
      existing = new CheckinEntity();
      existing.habitId = habitId;
      existing.userId = userId;
      existing.date = date;
      existing.persist();
    }

    existing.done = asBoolean(payload.get("done"), false);
    existing.count = Math.max(1, asInt(payload.get("count"), 1));
    existing.version = Math.max(existing.version, asInt(payload.get("version"), 0)) + 1;
    existing.updatedAt = nextSyncDate(clientUpdated, existing.updatedAt);

    applied.add(op.id());
  }

  private void deleteCheckin(
      String userId,
      String habitId,
      LocalDate date,
      Map<String, Object> payload,
      CheckinEntity existing
  ) {
    var tombstone = new TombstoneEntity();
    tombstone.userId = userId;
    tombstone.entity = "checkin";
    var payloadId = asString(payload.get("id"));
    tombstone.entityId = payloadId != null ? payloadId
        : (existing != null ? existing.id : habitId + ":" + date);
    tombstone.version = asInt(payload.get("version"), 1);
    tombstone.deletedAt = nextSyncDate(parseInstantOrNow(asString(payload.get("updatedAt"))));
    tombstone.persist();

    CheckinEntity.delete("habitId = ?1 and userId = ?2 and date = ?3", habitId, userId, date);
  }

  // ─── Deduplication log ────────────────────────────────────────────────────

  /** Atomically inserts the opId. Returns true if inserted (new op), false if duplicate. */
  private boolean tryCreateLog(String opId) {
    try {
      SyncOpLogEntity log = new SyncOpLogEntity();
      log.opId = opId;
      log.persistAndFlush();
      return true;
    } catch (PersistenceException ex) {
      // Duplicate opId and any other persistence error are treated as not-inserted.
      return false;
    }
  }

  // ─── Serialization ────────────────────────────────────────────────────────

  private SyncDtos.HabitDto serializeHabit(HabitEntity habit) {
    return new SyncDtos.HabitDto(
        habit.id,
        habit.name,
        habit.description == null ? "" : habit.description,
        habit.color,
        habit.icon,
        habit.frequency,
        parseJsonOrNull(habit.customDays),
        parseJsonOrNull(habit.schedule),
        habit.targetStreak,
        habit.dailyTarget,
        parseJsonOrEmptyList(habit.tags),
        habit.archived,
        toSyncIso(habit.createdAt),
        toSyncIso(habit.updatedAt),
        habit.version,
        habit.sortOrder == null ? 0 : habit.sortOrder.intValue(),
        habit.reminderTime,
        habit.reminderEnabled,
        habit.type,
        parseJsonOrEmptyList(habit.freezeDays)
    );
  }

  private SyncDtos.CheckinDto serializeCheckin(CheckinEntity checkin) {
    return new SyncDtos.CheckinDto(
        checkin.id,
        checkin.habitId,
        checkin.date.toString(),
        checkin.done,
        checkin.count,
        toSyncIso(checkin.updatedAt),
        checkin.version
    );
  }

  private SyncDtos.TombstoneDto serializeTombstone(TombstoneEntity tombstone) {
    return new SyncDtos.TombstoneDto(
        tombstone.id,
        tombstone.entity,
        tombstone.entityId,
        toSyncIso(tombstone.deletedAt),
        tombstone.version
    );
  }

  // ─── Cursor helpers ───────────────────────────────────────────────────────

  private CursorRow parseCursor(String raw) {
    if (raw == null || raw.isBlank()) {
      return null;
    }
    try {
      var data = objectMapper.readValue(raw, MAP_TYPE);
      var updatedAtStr = asString(data.get("updatedAt"));
      var id = asString(data.get("id"));
      if (updatedAtStr == null || id == null) {
        return null;
      }
      return new CursorRow(Instant.parse(updatedAtStr), id);
    } catch (Exception ex) {
      return null;
    }
  }

  private String calculateNextCursor(List<CursorRow> rows) {
    CursorRow latest;
    if (rows.isEmpty()) {
      // Always return a cursor so clients know sync happened up to now
      latest = new CursorRow(Instant.now(), "");
    } else {
      latest = rows.stream().max((a, b) -> {
        var cmp = a.updatedAt().compareTo(b.updatedAt());
        return cmp != 0 ? cmp : a.id().compareTo(b.id());
      }).orElse(new CursorRow(Instant.now(), ""));
    }

    try {
      return objectMapper.writeValueAsString(Map.of(
          "updatedAt", toSyncIso(latest.updatedAt()),
          "id", latest.id()
      ));
    } catch (Exception ex) {
      return null;
    }
  }

  // ─── Normalization helpers ────────────────────────────────────────────────

  /** Validates HH:MM format, returns null if invalid (mirrors NestJS normalizeReminderTime). */
  private String normalizeReminderTime(String value) {
    if (value == null || value.isBlank()) {
      return null;
    }
    if (!value.matches("^\\d{2}:\\d{2}$")) {
      return null;
    }
    var parts = value.split(":");
    int hours = Integer.parseInt(parts[0]);
    int minutes = Integer.parseInt(parts[1]);
    if (hours < 0 || hours > 23 || minutes < 0 || minutes > 59) {
      return null;
    }
    return String.format("%02d:%02d", hours, minutes);
  }

  /** Filters customDays to valid values 0–6, deduplicates (mirrors NestJS normalizeCustomDays). */
  @SuppressWarnings("unchecked")
  private String normalizeCustomDaysJson(Object value) {
    if (!(value instanceof List)) {
      return null;
    }
    var seen = new HashSet<Integer>();
    var result = new ArrayList<Integer>();
    for (var item : (List<?>) value) {
      if (item instanceof Number num) {
        int day = (int) num.doubleValue();
        if (day >= 0 && day <= 6 && seen.add(day)) {
          result.add(day);
        }
      }
    }
    if (result.isEmpty()) {
      return null;
    }
    return jsonOrNull(result);
  }

  /**
   * Validates freezeDays — each entry must match yyyy-MM-dd; deduplicates and sorts.
   * Falls back to existing if payload is absent/invalid.
   */
  @SuppressWarnings("unchecked")
  private String normalizeFreezeDaysJson(Object payloadValue, String existing) {
    if (payloadValue == null) {
      return existing != null ? existing : "[]";
    }
    if (!(payloadValue instanceof List)) {
      return "[]";
    }
    var seen = new TreeSet<String>();
    for (var item : (List<?>) payloadValue) {
      if (item instanceof String s && s.matches("^\\d{4}-\\d{2}-\\d{2}$")) {
        seen.add(s);
      }
    }
    return jsonOrNull(new ArrayList<>(seen));
  }

  /** Mirrors NestJS resolveSortOrder — falls back to existing when payload is invalid. */
  private BigInteger resolveSortOrder(Object payload, BigInteger existing) {
    if (payload instanceof Number num) {
      double val = num.doubleValue();
      if (Double.isFinite(val)) {
        return BigInteger.valueOf((long) val);
      }
    }
    return existing != null ? existing : BigInteger.ZERO;
  }

  /** Mirrors NestJS resolveDailyTarget — min 1, falls back to existing. */
  private int resolveDailyTarget(Object payload, int existingValue) {
    if (payload instanceof Number num) {
      double val = num.doubleValue();
      if (Double.isFinite(val)) {
        return Math.max(1, (int) val);
      }
    }
    return Math.max(1, existingValue > 0 ? existingValue : 1);
  }

  // ─── Utility helpers ──────────────────────────────────────────────────────

  private Instant normalizeInstant(String value) {
    return value != null ? parseInstantOrNow(value) : Instant.now();
  }

  private Instant parseInstantOrNow(String value) {
    if (value == null || value.isBlank()) {
      return Instant.now();
    }
    try {
      return Instant.parse(value);
    } catch (Exception ex) {
      return Instant.now();
    }
  }

  private LocalDate toLocalDate(String value) {
    try {
      if (value != null && value.length() >= 10) {
        return LocalDate.parse(value.substring(0, 10));
      }
    } catch (Exception ignored) {
    }
    return LocalDate.now(ZoneOffset.UTC);
  }

  private Instant nextSyncDate(Instant... values) {
    var max = Instant.now();
    for (var value : values) {
      if (value != null && value.isAfter(max)) {
        max = value;
      }
    }
    return max.plusSeconds(1);
  }

  private String toSyncIso(Instant instant) {
    if (instant == null) {
      return ISO.format(OffsetDateTime.now(ZoneOffset.UTC));
    }
    return ISO.format(OffsetDateTime.ofInstant(instant, ZoneOffset.UTC));
  }

  private Map<String, Object> toMap(Map<String, Object> payload) {
    return payload == null ? new HashMap<>() : payload;
  }

  private String asString(Object value) {
    if (value == null) {
      return null;
    }
    var text = String.valueOf(value);
    return text.isBlank() ? null : text;
  }

  private String nullableString(Object value) {
    return value == null ? null : String.valueOf(value);
  }

  private int asInt(Object value, int fallback) {
    if (value instanceof Number number) {
      return number.intValue();
    }
    if (value instanceof String text) {
      try {
        return Integer.parseInt(text);
      } catch (Exception ignored) {
      }
    }
    return fallback;
  }

  private boolean asBoolean(Object value, boolean fallback) {
    if (value instanceof Boolean bool) {
      return bool;
    }
    if (value instanceof String text) {
      return "true".equalsIgnoreCase(text) || "1".equals(text);
    }
    return fallback;
  }

  private String normalizeType(String type) {
    return "negative".equals(type) ? "negative" : "positive";
  }

  private String jsonOrNull(Object value) {
    if (value == null) {
      return null;
    }
    try {
      return objectMapper.writeValueAsString(value);
    } catch (Exception ex) {
      return null;
    }
  }

  private Object parseJsonOrNull(String json) {
    if (json == null || json.isBlank()) {
      return null;
    }
    try {
      return objectMapper.readValue(json, Object.class);
    } catch (Exception ex) {
      return null;
    }
  }

  private Object parseJsonOrEmptyList(String json) {
    var parsed = parseJsonOrNull(json);
    return parsed == null ? List.of() : parsed;
  }

  private record CursorRow(Instant updatedAt, String id) {
  }
}
